File: ando_library/src/main/java/com/ando/library/views/AbstractDoubleClickListener.java
package com.ando.library.views;

import android.view.View;

import com.ando.toolkit.L;

/**
 * Title: AbstractDoubleClickListener
 * <p>
 * Description: 双击监听 & 双击防抖
 * </p>
 *
 * @author javakam
 * @date 2020-04-15
 */
public abstract class AbstractDoubleClickListener implements View.OnClickListener {

    //默认双击事件响应间隔 300 毫秒
    private static final long DOUBLE_RESPONSE_TIME = 300;
    //双击防抖 1500 毫秒
    private static final long SHAKE_DURATION = 1500;

    private long lastClickTime;
    private long doubleRespDuration;
    private long shakeDuration;

    public AbstractDoubleClickListener() {
        this(DOUBLE_RESPONSE_TIME, SHAKE_DURATION);
    }

    public AbstractDoubleClickListener(long doubleRespDuration, long shakeDuration) {
        this.doubleRespDuration = doubleRespDuration;
        this.shakeDuration = shakeDuration;
    }

    @Override
    public void onClick(View v) {
        long currentTimeMillis = System.currentTimeMillis();
        long diffTimeMillis = currentTimeMillis - lastClickTime;

        if (diffTimeMillis < doubleRespDuration) {
            L.w("time","time 222");
            onDoubleClick(v);
        }

        lastClickTime = currentTimeMillis;

        // > shakeDuration 防抖: 需要大于这个时间差才可以再次响应时间
//        if (diffTimeMillis > shakeDuration) {
//            L.w("time","time 1111");
//        }


    }

    public abstract void onDoubleClick(View v);

}

File: ando_library/src/main/java/com/ando/library/views/AbstractDoubleClickListener.kt
package com.ando.library.views

import android.view.View
import com.ando.toolkit.log.L

/**
 * Title: AbstractDoubleClickListener
 *
 * Description: 双击监听 & 双击防抖
 *
 * @author javakam
 * @date 2020-04-15
 */
abstract class AbstractDoubleClickListener constructor(
    private val doubleRespDuration: Long = DOUBLE_RESPONSE_TIME,
    private val shakeDuration: Long = SHAKE_DURATION
) : View.OnClickListener {
    
    private var lastClickTime: Long = 0
    
    override fun onClick(v: View) {
        val currentTimeMillis = System.currentTimeMillis()
        val diffTimeMillis = currentTimeMillis - lastClickTime
        if (diffTimeMillis < doubleRespDuration) {
            L.w("time", "time 222")
            onDoubleClick(v)
        }
        lastClickTime = currentTimeMillis

//      > shakeDuration 防抖: 需要大于这个时间差才可以再次响应时间
//      if (diffTimeMillis > shakeDuration) {
//          L.w("time","time 1111");
//      }
    }

    abstract fun onDoubleClick(v: View?)

    companion object {
        //默认双击事件响应间隔 300 毫秒
        private const val DOUBLE_RESPONSE_TIME: Long = 300

        //双击防抖 1500 毫秒
        private const val SHAKE_DURATION: Long = 1500
    }
}